felix/bpf/tc/attach.go

```diff
@@ -96,43 +96,56 @@ func (ap *AttachPoint) AlreadyAttached(object string) (int, bool) {
 	return -1, false
 }
 
-// AttachProgram attaches a BPF program from a file to the TC attach point
-func (ap *AttachPoint) AttachProgram() (int, error) {
-	logCxt := log.WithField("attachPoint", ap)
-
-	filename := ap.FileName()
-	binaryToLoad := path.Join(bpf.ObjectDir, filename)
-
-	progsToClean, err := ap.listAttachedPrograms()
-	if err != nil {
-		return -1, err
-	}
-	obj, err := libbpf.OpenObject(binaryToLoad)
+func (ap *AttachPoint) loadObject(ipVer int, file string) (*libbpf.Obj, error) {
+	obj, err := libbpf.OpenObject(file)
 	if err != nil {
-		return -1, err
+		return nil, err
 	}
-	defer obj.Close()
 
 	for m, err := obj.FirstMap(); m != nil && err == nil; m, err = m.NextMap() {
 		// In case of global variables, libbpf creates an internal map <prog_name>.rodata
 		// The values are read only for the BPF programs, but can be set to a value from
 		// userspace before the program is loaded.
 		if m.IsMapInternal() {
 			if err := ap.ConfigureProgram(m); err != nil {
-				return -1, fmt.Errorf("failed to configure %s: %w", filename, err)
+				return nil, fmt.Errorf("failed to configure %s: %w", file, err)
 			}
 			continue
 		}
 
 		if err := ap.setMapSize(m); err != nil {
-			return -1, fmt.Errorf("error setting map size %s : %w", m.Name(), err)
+			return nil, fmt.Errorf("error setting map size %s : %w", m.Name(), err)
 		}
 		pinPath := bpf.MapPinPath(m.Type(), m.Name(), ap.Iface, ap.Hook)
 		if err := m.SetPinPath(pinPath); err != nil {
-			return -1, fmt.Errorf("error pinning map %s: %w", m.Name(), err)
+			return nil, fmt.Errorf("error pinning map %s: %w", m.Name(), err)
 		}
 	}
 
+	if err := obj.Load(); err != nil {
+		return nil, fmt.Errorf("error loading program: %w", err)
+	}
+
+	err = ap.updateJumpMap(ipVer, obj)
+	if err != nil {
+		return nil, fmt.Errorf("error updating jump map %v", err)
+	}
+
+	return obj, nil
+}
+
+// AttachProgram attaches a BPF program from a file to the TC attach point
+func (ap *AttachPoint) AttachProgram() (int, error) {
+	logCxt := log.WithField("attachPoint", ap)
```

Comment:
I think it's probably better to use a common method with ip version argument to not repeat the same code here. 
For example, `RememberAttachedProg` and other related functions save the name of object file and its hash to prevent loading an object if not changed. With the new approach, it only save the name of v4 object file and reloads v6 object every time.

I think to not miss other logic, it's probably better to use a loop and call a common function to take care of everything.

Reply:
If I understand you concern right, it is true that reloading the v6 objects is broken. However, I left it so intentionally because (a) it is never used in production (and now neither in fv tests) and (b) this is all going to change so I did not want to spend much time on it. What will remain is the split which I needed to prototype :)